# Predicting Interactions of m6A regulated RNA-Binding Proteins

## Data Processing
### PAR-CLIP Data
For the protein-RNA interactions we used the pre-processed PAR-CLIP dataset from Mukherjee et al. 2019, which analyzed 66 proteins in HEK293 cells. The methylation data came from a miCLIP dataset that has been also done on HEK293 cells.
### MI-CLIP Data
Available in data folder.

### Generate Positive Samples
As a fist step we need to generate positive and negative samples, that we can use as classified inputs to train our model on.
The positive samples are available in the parclip dataset but we have to extend them to generate sequences with a length of 200 amino acids (AA) as the length of the original bed files would vary.
```
$ cat proteinFile.bed | awk '$3 = $3 + 100 - sprintf("%.0f", (($3 - $2)/2)), $2 = $3 - 200' \
> positives.bed
```
Command to replace the spaces with tabs (Bedtools standard).
```
$ sed 's/ /\t/g' positives.bed > positives.bed
```
Dynamically determine the center of the sequence and extend the length from that position in both direction with a length of 100.
This ensures that the original sequence is centered around the chromStart-chromEnd.

### Generate Negative Samples
Negative samples of type 1 for the model are generated by sampling randomly in a window around the positives e.g. at distance 200nt from both left and from right.</br>
From the right:
```
$ cat proteinFile.bed | bedtools slop -i stdin -g XPO5 -l -250 -r 250 | \
awk '$3 = $3 + 100 - sprintf("%.0f", (($3 - $2)/2)), $2 = $3 - 200' \
> negativesType1Right.bed
```
```
$sed 's/ /\t/g' negativesType1Right.bed > negativesType1Right.bed
```
From the left:
```
$ cat proteinFile.bed | bedtools slop -i stdin -g XPO5 -l 250 -r -250 | \
awk '$3 = $3 + 100 - sprintf("%.0f", (($3 - $2)/2)), $2 = $3 - 200' \
>  negativesType1Left.bed
```
```
$ sed 's/ /\t/g' negativesType1Left.bed > negativesType1Left.bed
```
Negative samples of type 2 are positives of other RBPs.
```
$ cat proteinFile.bed | awk '$3 = $3 + 100 - sprintf("%.0f", (($3 - $2)/2)), $2 = $3 - 200' \
>  negativesType2.bed
```
```
$ sed 's/ /\t/g' negativesType2.bed > negativesType2.bed
```
### Genome Data
```
$ wget https://ftp.ebi.ac.uk/pub/databases/gencode/Gencode_human/release_40/GRCh38.primary_assembly.genome.fa.gz
$ gzip -d GRCh38.primary_assembly.genome.fa.gz
$ bedtools getfasta -fi GRCh38.primary_assembly.genome.fa -bed positives.bed > positives.fasta
```
```
$ cat -n positives.fasta | sort -uk2 | sort -nk1 | cut -f2- > positives.fasta
```
Remove duplicate lines from a file assuming you don't mind that lines are sorted.
### Methylation sites
```
$ bedtools intersect -a parclip_data/positives.bed \
-b miclip_data/GSE63753_hek293.abcam.CIMS.m6A.9536.bed \
> intersections.bed
```

### Generate Data-Set out of MI-CLIP Data and processed PAR-CLIP
Generate labeled and encoded sequence-data out of modified positives, negatives and different encodings.
```
$ python bedToFastaPreprocessing.py miclip.bed positives.fasta "positives" "methylationRate" 0
```
