{
 "cells": [
  {
   "cell_type": "code",
   "execution_count": 33,
   "metadata": {},
   "outputs": [],
   "source": [
    "def getUniqueMethylationSitesFromIntersectionsFileOfProtein(file_path) -> set:\n",
    "    # file with intersection with the protein needs to be in the same directory\n",
    "    file = open(file_path, 'r')\n",
    "    lines = file.readlines()\n",
    "    \n",
    "    uniqueMethylationSites = set()\n",
    "    for line in lines:\n",
    "        line = line.split(\"\\t\")\n",
    "        uniqueMethylationSites.add((str(line[0]), str(line[1]),str(line[8]).rstrip()))\n",
    "    \n",
    "    file.close()\n",
    "    return uniqueMethylationSites\n",
    "\n",
    "def getStatusCode(status) -> int:\n",
    "    if status == \"positive\":\n",
    "        return 1\n",
    "    elif status == \"negative\":\n",
    "        return 0\n",
    "    else:\n",
    "        raise ValueError('Status-Code for protein-label not *positive* or *negative*. Wrong format.')\n",
    "        \n",
    "def getEncodingCode(encoding) -> str:\n",
    "    if encoding == \"oneHot\":\n",
    "        return 0\n",
    "    elif encoding == \"methylationRate\":\n",
    "        return 1\n",
    "    else:\n",
    "        raise ValueError('Encoding-Code not *oneHot* or *methylationRate*. Wrong format.')\n",
    "        \n",
    "def createProteinSequenceEncodingAnnotation(file_path, status, encodingCode, methylationSites) -> None:\n",
    "    chromosome = \"\"\n",
    "    seqBeg = 0\n",
    "    seqEnd = 0\n",
    "    label = getStatusCode(status)\n",
    "    encodingCode = getEncodingCode(encodingCode)\n",
    "    methylationRate = 0\n",
    "    \n",
    "    encodedSequences = open(\"positives.fasta\", \"w\")\n",
    "    inputSequences = open(file_path, 'r')\n",
    "    lines = inputSequences.readlines()\n",
    "    \n",
    "    for line in lines:\n",
    "        if line[0] == \">\":\n",
    "            chromosome = line[1:].partition(\":\")[0].strip()\n",
    "            seqList = (line.partition(\":\")[2]).partition(\"-\")\n",
    "            seqBeg = int(seqList[0])\n",
    "            seqEnd = int(seqList[2])\n",
    "            encodedSequences.write(\">\" + str(label) + \"\\n\")\n",
    "            # print(\">\" + str(label), end=\"\\n\") - debugging\n",
    "        else:\n",
    "            encoding = \"0\" * 200\n",
    "            for methylationSite in methylationSites:\n",
    "                position = int(methylationSite[1])\n",
    "                if seqBeg <= position < seqEnd and chromosome == methylationSite[0]:\n",
    "                    if encodingCode == 0:\n",
    "                        encoding = encoding[:position-seqBeg] + \"1\" + encoding[position-seqBeg + 1:]\n",
    "                    elif encodingCode == 1:\n",
    "                        rate = str(int(float(\"{:.1f}\".format(float(methylationSite[2])))*10))\n",
    "                        encoding = encoding[:position-seqBeg] + rate + encoding[position-seqBeg + 1:]\n",
    "                        \n",
    "            encodedSequences.write(line + encoding + \"\\n\")    \n",
    "            # print(line + encoding, end=\"\\n\") - debugging\n",
    "    \n",
    "    encodedSequences.close()\n",
    "    inputSequences.close()\n",
    "    return"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": [
    "if __name__ == '__main__':\n",
    "    path = 'GSE63753_hek293.abcam.CIMS.m6A.9536.bed'\n",
    "    s = getUniqueMethylationSitesFromIntersectionsFileOfProtein(path)\n",
    "\n",
    "    path2 = \"SRR488727_positive_extended_fasta_unique - Kopie\"\n",
    "    createProteinSequenceEncodingAnnotation(path2, \"positive\", \"methylationRate\", s)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3",
   "language": "python",
   "name": "python3"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 4
}
